DR-2893 Make the Job Modal URL Addressable

nmalfroy opened the pull request:
This PR does a few things:
- It makes it so that a query parameter can be used to make a job description modal open (via `?expandedJob=<jobId>`)
- This is nice because now you can refresh the page and the job info is still there
- Note: the job doesn't have to be on the page to load
- It also allows linking from a failed job.  This PR also adds the ability to do that.

Here's the interaction.  Note the url

![jobmodal](https://user-images.githubusercontent.com/5633787/219502655-06b3f6c0-9aa3-4105-9184-1ed1f27c44d5.gif)


Note: newer designs for the error toast:
**Plain**
![image](https://user-images.githubusercontent.com/5633787/222773857-68ffc135-3e97-41f1-be45-00beb57095b4.png)

**With a job link**
![image](https://user-images.githubusercontent.com/5633787/222773876-866d6e7e-b407-4994-98ae-0540d4c546a9.png)

**Handle large messages**
![image](https://user-images.githubusercontent.com/5633787/222774024-d788f2cd-7a29-4a93-85e9-ae1dfd8e5d6d.png)

You can see this in action at: 
https://jade-nm.datarepo-dev.broadinstitute.org

nmalfroy committed:
Add utility method to render url query parameters

nmalfroy committed:
Make which job modal is open url addressable

nmalfroy committed:
Typescriptify and add job link to notifications

nmalfroy committed:
Linting and such

okotsopoulos commented on src/components/JobView.test.tsx:
```diff
@@ -44,19 +45,22 @@ const initialState = {
 
 beforeEach(() => {
-  const mockStore = createMockStore([]);
+  // sinon.replace(history, 'push', (path: any) => console.log('!!!!!', path));
```
I think these were left in from debugging!

nmalfroy replied:
derp.  good catch!

okotsopoulos commented on src/components/table/JobTable.tsx:
```diff
@@ -45,17 +47,23 @@ interface IProps extends WithStyles<typeof styles> {
       failed: { icon: `fa fa-circle-xmark ${classes.statusIconFailed}`, label: 'Failed' },
     };
 
+    // If the expanded job is not on the page, add the modal explicitly at the top level
```
I appreciate the comment here!

okotsopoulos commented on src/modules/notifications.tsx:
```diff
@@ -17,14 +30,14 @@ export function showNotification(err) {
     dismiss: {
       duration: 5000,
-      touch: false,
+      touch: true,
```
Want to confirm the expected behavior here: I'm seeing that all toast notifications now must be clicked on to close, including those without associated jobIds.

I think I like this better, we've heard from users that their toast disappeared before they could save the information.  But the behavior is not 100% clear.  Would it be possible to give toasts an "x" icon that you click to close?  That may be OOS.

nmalfroy replied:
Yeah, I was trying to style it and it was driving me nuts so I just left it on.  I think it's better this way.  I can certainly style...I want to sync with Lou about the job link anyways

snf2ye replied:
This is an interesting change - I could imagine a user trying to copy/paste the error message, but that would actually dismiss the message, right? I think an "x" icon would be a good idea.

okotsopoulos commented on src/components/Toast.tsx:
```diff
@@ -0,0 +1,79 @@
+      </CardContent>
+      {jobId && (
+        <CardContent className={classes.jobLinkContainer}>
+          <Link to={`/activity?expandedJob=${jobId}`}>Click to view details</Link>
```
What do you think about having this open out in a new tab?  For both snapshot creation and dataset creation, I could see it being frustrating to the user to lose their state on clicking.  Even better would be to not need to leave the page at all, if we could have the same job details modal from the activity page pop up here without going to the activity page itself.

I also wonder about the text of this link, I wonder if it would be helpful to expose the job ID here instead.

nmalfroy replied:
Oh interesting...yeah, I think the modal should work.  I'll check in with Lou.  My only fear is getting into modal overload

okotsopoulos reviewed (approved):
I have a feeling that direct links to TDR jobs in the UI will make our users very happy!

I left a few suggestions in the comments, but nothing major enough to hold my approval.

nmalfroy committed:
remove logging code

snf2ye commented on src/components/job/JobResultModal.tsx:
```diff
@@ -75,22 +81,26 @@ function JobResultModal({
       dispatch(getJobResult({ id }));
     }
-  }, [dispatch, seeMore, id]);
+  }, [dispatch, expandedJob, id]);
```
(Just a thought, I'm good with it as is!)
If we set the seeMore variable here, we could avoid the copied logic `expandedJob === id` and instead use the seeMore Variable on[ L125](https://github.com/DataBiosphere/jade-data-repo-ui/pull/1288/commits/85ea5b63c945fe15483eba7b9943988140d96269#diff-d36e518aa80e39d550b639fd716ed3e555ab7425c330083a7aefd35952f74da4R125). 
```
const [seeMore, setSeeMore] = useState(false);
  useEffect(() => {
    setSeeMore(expandedJob === id);
    if (seeMore) {
      dispatch(getJobResult({ id }));
    }
  }, [dispatch, expandedJob, id, seeMore]);
```

nmalfroy committed:
Notification improvements:

- Job results can be viewed from anywhere in the UI (via the url parameter)
  - and they even show all of the job info :)
- Error notification was restyled
- It not has an explicit close button (clicking the dialog still works too)
- Fix issue where resubmitting a snapshot was not working if resubmitting (introduced in this PR so all good)

nmalfroy committed:
Fix tests and remove auto-redirect on snapshot create

nmalfroy committed:
Drive by: fix regions to be consistent

nmalfroy committed:
Add right padding for text in Toast to not overlap with close button

okotsopoulos reviewed (approved):
Thanks so much for the additional work here on error modal behavior.  I love it!

I tried out a snapshot creation job that I knew would fail to induce an error modal with job details.  Keeping everything on the same page in this case made it easy to correct my request error and resubmit without having to reenter the request :)

nmalfroy committed:
Have to click the 'x' to close the Toast

snf2ye reviewed (approved):
These changes look great, thank you!
